modules/ip/datapath/arp_input.c

```diff
@@ -17,125 +17,75 @@
 enum {
 	OP_REQUEST = 0,
 	OP_REPLY,
+	CONTROL,
 	OP_UNSUPP,
 	PROTO_UNSUPP,
 	ERROR,
 	DROP,
-	IP_OUTPUT,
 	EDGE_COUNT,
 };
 
-static inline void update_nexthop(
-	struct rte_graph *graph,
-	struct rte_node *node,
-	struct nexthop *nh,
-	uint64_t now,
-	uint16_t iface_id,
-	const struct rte_arp_hdr *arp
-) {
-	struct ip_output_mbuf_data *o;
-	struct rte_mbuf *m, *next;
-
-	// Static next hops never need updating.
-	if (nh->flags & GR_NH_F_STATIC)
-		return;
-
-	rte_spinlock_lock(&nh->lock);
-
-	// Refresh all fields.
-	nh->last_reply = now;
-	nh->iface_id = iface_id;
-	nh->flags |= GR_NH_F_REACHABLE;
-	nh->flags &= ~(GR_NH_F_STALE | GR_NH_F_PENDING | GR_NH_F_FAILED);
-	nh->ucast_probes = 0;
-	nh->bcast_probes = 0;
-	nh->lladdr = arp->arp_data.arp_sha;
-
-	// Flush all held packets.
-	m = nh->held_pkts_head;
-	while (m != NULL) {
-		next = queue_mbuf_data(m)->next;
-		o = ip_output_mbuf_data(m);
-		o->nh = nh;
-		o->iface = NULL;
-		rte_node_enqueue_x1(graph, node, IP_OUTPUT, m);
-		m = next;
-	}
-	nh->held_pkts_head = NULL;
-	nh->held_pkts_tail = NULL;
-	nh->held_pkts_num = 0;
-
-	rte_spinlock_unlock(&nh->lock);
-}
-
 static uint16_t
 arp_input_process(struct rte_graph *graph, struct rte_node *node, void **objs, uint16_t nb_objs) {
 	struct nexthop *remote, *local;
-	struct arp_mbuf_data *arp_data;
+	struct rte_arp_hdr *arp, *t;
 	const struct iface *iface;
-	struct rte_arp_hdr *arp;
 	struct rte_mbuf *mbuf;
 	rte_edge_t edge;
-	ip4_addr_t sip;
-	uint64_t now;
-
-	now = rte_get_tsc_cycles();
 
 	for (uint16_t i = 0; i < nb_objs; i++) {
 		mbuf = objs[i];
 
 		// ARP protocol sanity checks.
 		arp = rte_pktmbuf_mtod(mbuf, struct rte_arp_hdr *);
-		if (rte_be_to_cpu_16(arp->arp_hardware) != RTE_ARP_HRD_ETHER) {
+		if (arp->arp_hardware != RTE_BE16(RTE_ARP_HRD_ETHER)) {
 			edge = PROTO_UNSUPP;
 			goto next;
 		}
-		if (rte_be_to_cpu_16(arp->arp_protocol) != RTE_ETHER_TYPE_IPV4) {
+		if (arp->arp_protocol != RTE_BE16(RTE_ETHER_TYPE_IPV4)) {
 			edge = PROTO_UNSUPP;
 			goto next;
 		}
-		switch (rte_be_to_cpu_16(arp->arp_opcode)) {
-		case RTE_ARP_OP_REQUEST:
+		switch (arp->arp_opcode) {
+		case RTE_BE16(RTE_ARP_OP_REQUEST):
 			edge = OP_REQUEST;
 			break;
-		case RTE_ARP_OP_REPLY:
+		case RTE_BE16(RTE_ARP_OP_REPLY):
 			edge = OP_REPLY;
 			break;
 		default:
 			edge = OP_UNSUPP;
 			goto next;
 		}
 
-		sip = arp->arp_data.arp_sip;
-		iface = eth_input_mbuf_data(mbuf)->iface;
-		local = ip4_addr_get_preferred(iface->id, sip);
```

Comment:
It seems the behavior around lookups is changed in this commit at the same time the moving from dp to cp is done. Please split unrelated changes in separate patches.

Reply:
done